### PR TITLE
Fix KCVSLog not correctly shutdowns after closure if storage is shutdowned shortly after closure

### PR DESCRIPTION
**Pull request by porunov**
In case we shutdown storage backend shortly after closure message pullers will not be closed and will be retried even when the thread is interrupted. It will log 'Could not read messages for timestamp [2021-08-18T20:29:59.057Z] (this read will be retried)'

Not sure yet how to test it properly

Signed-off-by: Oleksandr Porunov <alexandr.porunov@gmail.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/log/kcvs/KCVSLog.java [hunk 1]
```diff
@@ -337,9 +337,23 @@ public synchronized void close() throws BackendException {
                 }
             }
         }
-        writeSetting(manager.senderId, MESSAGE_COUNTER_COLUMN, numMsgCounter.get());
-        store.close();
-        manager.closedLog(this);
+        try {
+            writeSetting(manager.senderId, MESSAGE_COUNTER_COLUMN, numMsgCounter.get());
+        } catch (Throwable e) {
+            log.error("Could not persist message counter [" + manager.senderId + "] ; message counter ["+numMsgCounter.get()+"]",e);
```
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/diskstorage/log/kcvs/KCVSLog.java:343]: We could avoid unnecessary string concatenation by using slf4j placeholders:
```suggestion
            log.error("Could not persist message counter [{}] ; message counter [{}]", manager.senderId, numMsgCounter.get(), e);
```

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/log/kcvs/KCVSLog.java [hunk 2]
```diff
@@ -777,6 +791,10 @@ public void run() {
                 if (e.getCause() instanceof PermanentBackendException) {
                     throw e;
                 }
+                if(Thread.currentThread().isInterrupted()){
+                    log.warn("Could not read messages for timestamp ["+messageTimeStart+"]. Thread was interrupted. Thus, no mo retries will be conducted.",e);
+                    throw e;
```
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/diskstorage/log/kcvs/KCVSLog.java:796]: Is it possible to test this?

**Review by li-boxuan — COMMENTED** [review]
Code logic looks good to me!

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/log/kcvs/KCVSLog.java (hunk 2 above)
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/log/kcvs/KCVSLog.java:796]: I will wait for CI to pass in #2757 and merge it. After that, there will be available a `TestLogger` in tests which we can use to intercept logs. I think we could than try to start backend, shutdown storage and close backend after that. I think in that case we will be able to see this log as KCVSLog will not be closed properly after storage shutdown (due to inability to communicate with the storage backend). I will try to reproduce it with Cassandra

Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/log/kcvs/KCVSLog.java:796]: After working on #2769 I couldn't reproduce this issue anymore. Most likely this was due to underlying storages improper shutdown.

**Comment by porunov** [comment]
Closing this PR in favor to #2769

**Pull request closed without merging**